### PR TITLE
Change folders and alternate state machine framework

### PR DESCRIPTION
**Pull request by mattsmart**
state_machine.py contains the classes for the simulation (Target, Domain, Genome)
genome_simulation.py may eventually perform the simulation

### CODE REVIEW

File: models/mathCAS/cut_modelling/state_machine.py [hunk 1]
```diff
@@ -0,0 +1,103 @@
...
+        self.sequence = sequence  # string, include PAM, should be ~ 23 chars, maybe need buffer on opposite end
+        self.start = start  # int
+        self.total_cuts = 0
+        self.cut_position = None  # number of nt from PAM site, starting at 0
+        self.repaired = True  # defined by open/closed
+        self.targetable = True  # defined by targetable or not (PAM broken or indel size > 5)
+        self.cut_probability = self.compute_cut_probability(dt)
+        self.shift = 0  # defined by sum of net indel sizes, used to compute frameshift if orf region
+        self.complex_concentration = complex_concentration  # conc of gRNA-cas9 complex inside nucleus
+
```
Comment (mattsmart) [inline models/mathCAS/cut_modelling/state_machine.py:28]: need sense field for direction of PAM and target

File: models/mathCAS/cut_modelling/state_machine.py [hunk 2]
```diff
@@ -0,0 +1,103 @@
...
+        # TODO: nucleotide insertion module
+        insert_left_nt = 'X'  # foo_nt_rand(insert_left)
+        insert_right_nt = 'Y'  # foo_nt_rand(insert_right)
+        # rewrite target sequence
+        sequence_left = self.sequence[0:self.cut_position-del_left] + insert_left_nt
+        sequence_right = insert_right_nt + self.sequence[self.cut_position+del_right:]
+        self.sequence = sequence_left + sequence_right  # may need to get buffer from genome data if deletion
+        self.cut_position = None
+        self.repaired = True
+        if net_indel_size > 5:  # or if PAM is broken
```
Comment (Ellmen) [inline models/mathCAS/cut_modelling/state_machine.py:63]: We might want to double-check that a broken PAM makes a site absolutely untargetable.

File: models/mathCAS/cut_modelling/state_machine.py [hunk 3]
```diff
@@ -0,0 +1,103 @@
...
+
+    def update_functionality(self):
+        if self.domain_type == "orf":
+            if not self.promoter.is_functional() or sum(target.get_shift() for target in self.targets) % 3 != 0:
+                self.functional = False
+            else:
+                self.functional = True
+        elif self.domain_type == "promoter":  # how to define functional promoter
+            self.functional = True
+        else:  # how to define functional NCR
```
Comment (Ellmen) [inline models/mathCAS/cut_modelling/state_machine.py:93]: Are we planning on worrying about this?

File: models/mathCAS/cut_modelling/state_machine.py [hunk 4]
```diff
@@ -0,0 +1,103 @@
...
+            self.functional = True
+
+
+class Genome(object):
+
+    def __init__(self, length, domains):
+        self.length = length  # int
+        self.sequence = None  # to be implemented
+        self.domains = domains  # list of all domains (ORFs, promoters, NCRs)
+        self.repaired = True
```
Comment (Ellmen) [inline models/mathCAS/cut_modelling/state_machine.py:103]: Looks good, we'll probably want to update the genome every time there is an indel at a target. Then we'll have to add a way of reindexing every target/domain after that site.

File: models/mathCAS/cut_modelling/state_machine.py (hunk 3 above)
Comment (mattsmart) [inline models/mathCAS/cut_modelling/state_machine.py:93]: not necessarily for an NCR target, but functional promoter is important for one of our big deletion scenarios

File: models/mathCAS/cut_modelling/state_machine.py (hunk 1 above)
Comment (mattsmart) [inline models/mathCAS/cut_modelling/state_machine.py:28]: also want buffer

Comment (mattsmart) [inline models/mathCAS/cut_modelling/state_machine.py:28]: also insertion left / right -> just insertion

**Pull request merged**